nixos/modules/services/desktop-managers/cosmic.nix

```
@@ -7,11 +7,39 @@
   config,
   lib,
   pkgs,
+  utils,
   ...
 }:
 
 let
   cfg = config.services.desktopManager.cosmic;
+  excludedCorePkgs = lib.lists.intersectLists corePkgs config.environment.cosmic.excludePackages;
+  # **ONLY ADD PACKAGES WITHOUT WHICH COSMIC CRASHES, NOTHING ELSE**
+  corePkgs =
+    with pkgs;
+    [
+      cosmic-applets
+      cosmic-applibrary
+      cosmic-bg
+      cosmic-comp
+      cosmic-files
+      config.services.displayManager.cosmic-greeter.package
+      cosmic-idle
+      cosmic-launcher
+      cosmic-notifications
+      cosmic-osd
+      cosmic-panel
+      cosmic-session
+      cosmic-settings
+      cosmic-settings-daemon
+      cosmic-workspaces-epoch
+    ]
+    ++ lib.optionals cfg.xwayland.enable [
+      # Why would you want to enable XWayland but exclude the package
+      # providing XWayland support? Doesn't make sense. Add `xwayland` to the
+      # `corePkgs` list.
+      xwayland
+    ];
```

`cosmic-applets` and `cosmic-settings` are not strictly needed by the COSMIC Desktop session.
Also, we could move the xwayland package addition outside this variable.

```suggestion
  corePkgs =
    with pkgs;
    [
      cosmic-applibrary
      cosmic-bg
      cosmic-comp
      cosmic-files
      config.services.displayManager.cosmic-greeter.package
      cosmic-idle
      cosmic-launcher
      cosmic-notifications
      cosmic-osd
      cosmic-panel
      cosmic-session
      cosmic-settings-daemon
      cosmic-workspaces-epoch
    ];
```

---

> I determined this list by running all existing `nixosTests.cosmic*` tests by removing one package at a time. When even one test failed, I added said removed package to the `corePkgs` list.

It probably fails because the test launches `cosmic-settings` and also launches an applet. But it is not required by the session, as you can see in: [pop-os/cosmic-session @ src/main.rs](https://github.com/pop-os/cosmic-session/blob/80209d5aadcfc05803eaa5b5099a5db1550d3fb1/src/main.rs), unlike the rest of the packages.

> About moving the `xwayland` package outside this list, please read the comment justifying it's inclusion. The `xwayland` package can be included or excluded using the `services.desktopManager.cosmic.xwayland.enable` option. It doesn't make sense if `services.desktopManager.cosmic.xwayland.enable` is enabled but the user is removing the `xwayland` package; given that the only thing that the `services.desktopManager.cosmic.xwayland.enable` option does is add or not add the `xwayland` package to `environment.systemPackages`.

I didn't mean to make xwayland excludable, I just think we can move it to outside the variable, to make the code cleaner

---

I agree with @HeitorAugustoLN but I think it really depends on how we define core packages.

For example, `cosmic-files` or even the greeter will are not needed for a stable cosmic session, right?

---

> I didn't mean to make xwayland excludable, I just think we can move it to outside the variable, to make the code cleaner

I don't follow. Move it where?